src/middleware/_utils.ts

```diff
@@ -1,30 +1,46 @@
 import type { IncomingMessage, OutgoingMessage } from "node:http";
 import type { ProxyServer } from "../server";
-import type { ProxyServerOptions } from "../types";
+import type {
+  NormalizedProxyTarget,
+  NormalizedProxyTargetUrl,
+  ProxyServerOptions,
+} from "../types";
+import type { Socket } from "node:net";
 
-export type ProxyMiddleware = (
+export type ResOfType<T extends "web" | "ws"> = T extends "ws"
+  ? T extends "web"
+    ? OutgoingMessage | Socket
+    : Socket
+  : T extends "web"
+    ? OutgoingMessage
+    : never;
+
+export type ProxyMiddleware<T extends OutgoingMessage | Socket> = (
   req: IncomingMessage,
-  res: OutgoingMessage,
-  opts: ProxyServerOptions & { target: URL; forward: URL },
-  server?: ProxyServer,
+  res: T,
```

Comment:
ws receives `net.Socket` instead of `http.OutgoingMessage` so many things depending on this needs to be generic.